### ISSUE TITLE
Not working in Safari with `mkdocs build` for repetitive searches

### ISSUE THREAD

**Issue by ronligt** [wilhelmer/mkdocs-localsearch#11]
When building the site-directory with `mkdocs build` the search-function only works the first time in Safari. After selecting a new page in the left-side menu search stops working and only displays `Initializing seach`.The console shows no error.

Also selecting one of the other pages found with search in the drop-down menu is not working.

## versions

mkdocs: 1.1.2
mkdocs-material: 6.0.1
mkdocs-localsearch: 0.7.0
iframe-worker 0.1.9
safari: 14.0

This was tested with a bare setup of MkDocs-Material (copying `index.html` for second page) and following the instructions in the README.md.

**Comment by wilhelmer** [comment]
Thanks for reporting, I can reproduce it. When navigating between pages using search or the nav, search fails. Not all the time, but very often. If I refresh the page via CMD-R, search always works again. 

Weird bug and probably again related to the iframe-worker. 

@squidfunk, if you could be so kind again ... but I have to say, this all seems to apply to Safari 14 only, which was just released and isn't even part of a final macOS release yet. So maybe we should't put too much energy in this.

**Comment by squidfunk** [comment]
Does it happen in Safari 13?

**Comment by ronligt** [comment]
I'm afraid I cannot test it on Safari 13. Since September 16, 2020 Safari 14 is the default version for macOS 10.14 Mojave and macOS 10.15 Catalina. After the latest macOS update version 14 replaced version 13 on my Mac (10.15.7).

**Comment by wilhelmer** [comment]
Oh okay, thought you had to download it seperately on Mojave and Catalina. But still, it's a .0 release.

**Comment by squidfunk** [comment]
It's wildly indeterministic. Just tested it on Safari and for me, most of the time it works, but sometimes it doesn't. It's probably related to the same problem you were trying to solve with the `promise_delay` option. Does it happen for you when setting `promise_delay`?

**Comment by wilhelmer** [comment]
Had the same thought, but that doesn't change anything. Tried `promise_delay: 100` and even `promise_delay: 500`.

**Comment by squidfunk** [comment]
I'll see what I can do when I find the time. It's either related to mkdocs-material or iframe-worker, so not a problem with this project in particular. For the time being, don't use Safari. It's the new IE 😉

**Comment by squidfunk** [comment]
I took some time of my lunch break and here are my preliminary findings:

- It's a Heisenbug! If you try to observe it, it alters its behavior: if the console is open, it doesn't occur. If it's closed, it occurs every 3-4th time for me
- It appears that in Safari, the `load` event is not fired consistently, so it appears to be a Safari bug. Thus, we need to find another way of detecting whether the `iframe` has loaded and is able to receive events via `postMessage`.

**Comment by squidfunk** [comment]
I was too curious again:

- First of all, I've fixed another bug where the order of multiple import scripts was indeterministic, leading to occasional errors, especially when using the Japanese language, as three files are being loaded.
- Second of all, here's an attempt to fix the problem with Safari. It's far from elegant, but I'd be interested whether it solves your problem and still works in all other browsers. If it does, we can refactor it and release it:

iframe-worker (custom build): [polyfill.zip](https://github.com/wilhelmer/mkdocs-localsearch/files/5318000/polyfill.zip)

**Comment by wilhelmer** [comment]
Yes, with your custom build, it works for me, in Safari 14 as well as Chrome and FF (latest versions). 🥳

**Comment by squidfunk** [comment]
Thanks for testing. I'll wrap up something this weekend, pushing out a new version of iframe-worker.

**Comment by squidfunk** [comment]
I think I've got a working solution. It seems that Safari will not execute the script injected into the `iframe` when visiting a page through navigation that was visited before in the same session. This is clearly a bug in Safari. To mitigate this behavior, I added cache busting in https://github.com/squidfunk/iframe-worker/commit/b90017545ea8940d0cec85b2ff5a0b59947a444e, which seems to do the trick, as my testing shows.

[iframe-worker-0.2.0]() contains those changes, plus the fixes for race conditions I mentioned earlier in https://github.com/wilhelmer/mkdocs-localsearch/issues/11#issuecomment-702714458
